# NAIVE BAYES

IMPORTAR DATOS


In [84]:
import pandas as pd

In [85]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [86]:
solicitantes = pd.read_excel('/content/drive/MyDrive/0. SolicitantesCrédito.xlsx')

In [87]:
solicitantes

Unnamed: 0  Edad    Sexo Estado civil  Hijos  Perscargo  Estrato  \
0              1    38      NR       Casado      1          0        3   
1              2    51      NR     Separado      6          1        4   
2              3    37   Mujer       Casado      2          0        3   
3              4    29      NR     Separado      1          0        2   
4              5    42   Mujer       Casado      3          4        2   
...          ...   ...     ...          ...    ...        ...      ...   
8373        8374    31  Hombre      Soltero      2          0        5   
8374        8375    38  Hombre     Separado      0          0        3   
8375        8376    43   Mujer       Casado      1          2        2   
8376        8377    37   Mujer       Casado      1          2        3   
8377        8378    24      NR      Soltero      0          0        2   

     Nivel de estudios  Tipo de contrato  Ingresos  ...  Plazo    Cuota (COP)  \
0            Bachiller           Ninguno   5424576  ...     48   82691.506204   
1              Técnico           Ninguno   1144064  ...     36   22781.212598   
2            Tecnólogo        Indefinido   2069853  ...     36   46294.596244   
3            Tecnólogo           Ninguno   1893080  ...     24   69564.950015   
4      Especializacion           Ninguno   3000367  ...     24  112478.307883   
...                ...               ...       ...  ...    ...            ...   
8373         Bachiller        Indefinido   5891088  ...     48  115108.056436   
8374         Bachiller  Labor Contratada   3092079  ...     36   96789.562091   
8375         Tecnólogo           Ninguno   2542007  ...     36   49713.608249   
8376         Tecnólogo        Indefinido   5677067  ...     36  163602.818587   
8377         Tecnólogo        Indefinido   1950392  ...     36   69024.928055   

                ConoCliente          Garantia            RespPat  \
0        Buenas Referencias      Sin Codeudor     Sin Patrimonio   
1        Buenas Referencias      Sin Codeudor  Prop.sin Hipoteca   
2        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
3        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
4        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
...                     ...               ...                ...   
8373  Buena Experiencia CFB      Sin Codeudor     Sin Patrimonio   
8374     Buenas Referencias      Sin Codeudor  Prop.sin Hipoteca   
8375     Buenas Referencias      Sin Codeudor     Sin Patrimonio   
8376     Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
8377     Buenas Referencias      Sin Codeudor     Sin Patrimonio   

                      MoraCome  PreApr  Prob.Default (PD)       Score  \
0     Con Experiencia sin Mora     1.0           0.041659  582.478943   
1     Con Experiencia sin Mora     1.0           0.234951  479.839966   
2     Con Experiencia sin Mora     0.0           0.170330  417.653107   
3     Con Experiencia sin Mora     1.0           0.201659  316.791107   
4     Con Experiencia sin Mora     1.0           0.140929  441.809570   
...                        ...     ...                ...         ...   
8373  Con Experiencia sin Mora     1.0           0.035120  659.471375   
8374  Con Experiencia sin Mora     1.0           0.101976  465.011536   
8375  Con Experiencia sin Mora     1.0           0.156707  423.237305   
8376  Con Experiencia sin Mora     1.0           0.037719  592.670837   
8377  Con Experiencia sin Mora     NaN                NaN         NaN   

           LGD  
0     0.063702  
1     0.465636  
2     0.290071  
3     0.278389  
4     0.203359  
...        ...  
8373  0.066962  
8374  0.165936  
8375  0.247763  
8376  0.051356  
8377       NaN  

[8378 rows x 35 columns]

PREPARAR VARIABLES

In [88]:
variables_independientes = [
    'Edad', 'Hijos', 'Perscargo', 'Estrato',
    'Ingresos', 'Egresos', 'Monto (EAD)', 'Plazo', 'Cuota (COP)'
]

print("\nVariables independientes seleccionadas:")
for i, var in enumerate(variables_independientes, 1):
    print(f"  {i}. {var}")

# Seleccionar variables independientes y dependiente
datos_modelo = solicitantes[variables_independientes + ['PreApr']]

print(f"\nRegistros antes de eliminar NaN: {len(datos_modelo)}")
datos_modelo = datos_modelo.dropna(subset=variables_independientes + ['PreApr'])
print(f"Registros después de eliminar NaN: {len(datos_modelo)}")

# Separar X e y
X_independientes = datos_modelo[variables_independientes]
y_dependiente = datos_modelo['PreApr']

print(f"\nDimensión X: {X_independientes.shape}")
print(f"Dimensión y: {y_dependiente.shape}")

# convertir a arrays

X_array = X_independientes.to_numpy()
y_array = y_dependiente.to_numpy()

print(f"\nX_array shape: {X_array.shape}")
print(f"y_array shape: {y_array.shape}")


Variables independientes seleccionadas:
  1. Edad
  2. Hijos
  3. Perscargo
  4. Estrato
  5. Ingresos
  6. Egresos
  7. Monto (EAD)
  8. Plazo
  9. Cuota (COP)

Registros antes de eliminar NaN: 8378
Registros después de eliminar NaN: 5842

Dimensión X: (5842, 9)
Dimensión y: (5842,)

X_array shape: (5842, 9)
y_array shape: (5842,)


ENTRENAR MODELO GAUSSIAN NAIVE BAYES

In [89]:
modelo_gnb = GaussianNB()
modelo_gnb.fit(X_array, y_array)

GaussianNB()

PREDICCIONES

In [90]:
y_predicho = modelo_gnb.predict(X_array)

MATRIZ DE CONFUSIÓN

In [91]:
matriz_confusion = confusion_matrix(y_array, y_predicho)
print("\nMatriz de Confusión:")
print(matriz_confusion)

# Extraer valores
verdaderos_negativos = matriz_confusion[0, 0]
falsos_positivos = matriz_confusion[0, 1]
falsos_negativos = matriz_confusion[1, 0]
verdaderos_positivos = matriz_confusion[1, 1]

print("\nValores de la Matriz de Confusión:")
print(f"  • Verdaderos Negativos (VN): {verdaderos_negativos}")
print(f"  • Falsos Positivos (FP):     {falsos_positivos}")
print(f"  • Falsos Negativos (FN):     {falsos_negativos}")
print(f"  • Verdaderos Positivos (VP): {verdaderos_positivos}")


Matriz de Confusión:
[[2539  420]
 [1041 1842]]

Valores de la Matriz de Confusión:
  • Verdaderos Negativos (VN): 2539
  • Falsos Positivos (FP):     420
  • Falsos Negativos (FN):     1041
  • Verdaderos Positivos (VP): 1842


METRICAS

In [92]:
exactitud = (verdaderos_positivos + verdaderos_negativos) / matriz_confusion.sum()
error = (falsos_positivos + falsos_negativos) / matriz_confusion.sum()
sensibilidad = verdaderos_positivos / (verdaderos_positivos + falsos_negativos)
precision = verdaderos_positivos / (verdaderos_positivos + falsos_positivos)
especificidad = verdaderos_negativos / (verdaderos_negativos + falsos_positivos)

print("\nMétricas del Modelo:")
print(f"  • Exactitud (Accuracy):   {exactitud:.2%}")
print(f"  • Error:                 {error:.2%}")
print(f"  • Sensibilidad:        {sensibilidad:.2%}")
print(f"  • Precisión:             {precision:.2%}")
print(f"  • Especificidad:        {especificidad:.2%}")



Métricas del Modelo:
  • Exactitud (Accuracy):   74.99%
  • Error:                 25.01%
  • Sensibilidad:        63.89%
  • Precisión:             81.43%
  • Especificidad:        85.81%


El modelo Gaussian Naive Bayes, entrenado con el conjunto extendido de 9 variables, presenta un desempeño global aceptable, alcanzando una exactitud del 74.99%, lo que indica que aproximadamente 3 de cada 4 solicitantes son clasificados correctamente. Este resultado sugiere que el modelo captura patrones relevantes en los datos, la precisión del 81.43% y la especificidad del 85.81% muestran que el modelo es confiable al identificar correctamente a los solicitantes no preaprobados y al reducir aprobaciones incorrectas.
Sin embargo, la sensibilidad del 63.89% muestra que el modelo no detecta a todos los solicitantes realmente preaprobables, presentando una tendencia conservadora en sus decisiones, esto es coherente con escenarios financieros donde el costo de aprobar erróneamente a un solicitante es mayor que el de rechazar a uno que sea potencialmente bueno.